File: sdks/go/idempotency_test.go
package metering

import (
	"regexp"
	"testing"
)

func TestGenerateIdempotencyKey_Deterministic(t *testing.T) {
	k1 := generateIdempotencyKey("cust_1", "api_calls", 1, "2026-03-21")
	k2 := generateIdempotencyKey("cust_1", "api_calls", 1, "2026-03-21")
	if k1 != k2 {
		t.Fatalf("expected deterministic keys, got %s != %s", k1, k2)
	}
}

func TestGenerateIdempotencyKey_DifferentInputs(t *testing.T) {
	k1 := generateIdempotencyKey("cust_1", "api_calls", 1, "2026-03-21")
	k2 := generateIdempotencyKey("cust_2", "api_calls", 1, "2026-03-21")
	if k1 == k2 {
		t.Fatal("expected different keys for different inputs")
	}
}

func TestGenerateIdempotencyKey_32HexChars(t *testing.T) {
	key := generateIdempotencyKey("cust_1", "metric", 5, "2026-01-01")
	if len(key) != 32 {
		t.Fatalf("expected 32 chars, got %d", len(key))
	}
	matched, _ := regexp.MatchString(`^[0-9a-f]{32}$`, key)
	if !matched {
		t.Fatalf("expected hex string, got %s", key)
	}
}

func TestGenerateRandomKey_Unique(t *testing.T) {
	k1 := generateRandomKey()
	k2 := generateRandomKey()
	if k1 == k2 {
		t.Fatal("expected unique random keys")
	}
}

File: aforo-metering-sdks/node/tests/idempotency.test.ts
import { generateIdempotencyKey, generateRandomKey } from '../src/idempotency';

describe('idempotency', () => {
  describe('generateIdempotencyKey', () => {
    it('should produce deterministic keys for same input', () => {
      const key1 = generateIdempotencyKey('cust_1', 'api_calls', 1, '2026-03-21T00:00:00Z');
      const key2 = generateIdempotencyKey('cust_1', 'api_calls', 1, '2026-03-21T00:00:00Z');
      expect(key1).toBe(key2);
    });

    it('should produce different keys for different inputs', () => {
      const key1 = generateIdempotencyKey('cust_1', 'api_calls', 1, '2026-03-21T00:00:00Z');
      const key2 = generateIdempotencyKey('cust_2', 'api_calls', 1, '2026-03-21T00:00:00Z');
      expect(key1).not.toBe(key2);
    });

    it('should produce 32-char hex string', () => {
      const key = generateIdempotencyKey('cust_1', 'metric', 5, '2026-01-01T00:00:00Z');
      expect(key).toHaveLength(32);
      expect(key).toMatch(/^[0-9a-f]{32}$/);
    });
  });

  describe('generateRandomKey', () => {
    it('should produce unique keys', () => {
      const key1 = generateRandomKey();
      const key2 = generateRandomKey();
      expect(key1).not.toBe(key2);
    });

    it('should produce UUID format', () => {
      const key = generateRandomKey();
      expect(key).toMatch(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/);
    });
  });
});
